### ISSUE TITLE
Strange behavior with `fs.watch` on file with JSON stringified -> first update isn't emitted if too fast after `fs.watch`

### ISSUE THREAD

**Issue by Apokalypt** [nodejs/help#4402]
### Node.js Version

v18.17.1

### NPM Version

9.6.7

### Operating System

23.2.0 Darwin

### Subsystem

fs

### Description

I would like to watch changes on a file storing JSON stringified data but the watcher created with `fs.watch` is not emitted if the change happen to quickly after the call of `fs.watch`. The strange thing is that if the content is only a text (without JSON format) the watcher emit the event correctly...

### Minimal Reproduction

```Javascript
const fs = require('fs');
const path = require('path');

const filePath = path.join(__dirname, 'test.json');
// Create the file if it doesn't exist
if (!fs.existsSync(filePath)) {
    fs.writeFileSync(filePath, "{ }");
}

fs.watch(filePath, (eventType, filename) => {
    console.log(`Event type is: ${eventType}`);
    if (filename) {
        console.log(`Filename provided: ${filename}`);
    } else {
        console.log('Filename not provided');
    }
});

fs.writeFileSync(filePath, "{ \"test\": \"test\", \"test2\": \"test2\" }"); // same with JSON.stringify(...)
// Listener above is called if we replace line above with:
// fs.writeFileSync(filePath, "very ok or not but I need a longer text to test something...");
```

### Expected
```
Event type is: change
Filename provided: test.json
```

### Received
*Nothing*

### Before You Submit

- [X] I have looked for issues that already exist before submitting this
- [X] My issue follows the guidelines in the README file, and follows the 'How to ask a good question' guide at https://stackoverflow.com/help/how-to-ask

**Comment by gireeshpunathil** [comment]
interesting!

able to reproduce with `v20.12.2` too.

if I replace the writing section with this code:

```js
if(process.argv[2] === 'json') {
	console.log('json')
	fs.writeFileSync(filePath, "{ \"test\": \"test\"}");
}
else {
	console.log('text')
	fs.writeFileSync(filePath, "hello!");
}
```

then the behaviour is arbitrary. sometimes the text change is called back, sometimes not. but the json callback is always omitted.

**Comment by Apokalypt** [comment]
Oh, I forgot to tell that I've also tested the behaviour on node version :
- 20.11.0
- 16.20.1

@gireeshpunathil Are you also on a MacOS computer or on another operating system ?

**Comment by gireeshpunathil** [comment]
yes - Mac . if I add an artificial delay using setTimeout before the write, things are fine; so looks like there is a race condition between the watch registration and the writing (though both are done by the same thread). not sure what is going on.

**Comment by Apokalypt** [comment]
@gireeshpunathil Do you think this could be considered as a bug and should be reported in the main repository containing the nodejs runtime source code?

**Comment by RedYetiDev** [comment]
> @gireeshpunathil Do you think this could be considered as a bug and should be reported in the main repository containing the nodejs runtime source code?

Maybe, it seems like a bug IMO
